CIP-plutus-modules/README.md

```diff
@@ -0,0 +1,1622 @@
+the `value scripts` variation.
+
+UPLC provides a fixpoint combinator, and this is how recursion is
+compiled. For the sake of argument, consider the well-known fixpoint
+combinator `Y` (in reality, `Y` is not suitable for use in a strict
+programming language, so the UPLC version is slightly different). We
+can imagine that a recursive function `f` is compiled as `Y h`, for
+some suitable `h`.
+
+The difficulty that arises is that `Y h` *is not a value*, and thus
+cannot appear at the top-level of a module, under the 'value script'
+restriction. It can be *normalised* into a value, of course, using
+```
+Y h ---> h (Y h)
+```
+and then reducing the application of `h`; this would need to be done
+by a compiler generating UPLC with the `value script`
+restriction. But reducing `h (Y h)` may well duplicate `Y h`. When
+this happens at CEK runtime it is not a problem, because all the
+occurrences of `Y h` are represented by the same pointer. But when the
+reductions are applied by a compiler, and the resulting term is
+serialized to UPLC code for inclusion in a script, then each
+occurrence of `Y h` will be serialized separately, losing sharing and
+causing code duplication in the resulting script. The result could be
+*larger* code, the opposite of what we are trying to achieve. Thus
+this method of compiling recursion fits badly with the 'value scripts'
+variation.
+
+Hence module-level recursion, which allows recursive occurrences of
+script values to be referred to via the `Self` variable instead of
+using a fixpoint combinator implemented in UPLC. To take advantage of
+this feature, the compiler will need to float occurrences of `fix`
+upwards, to the top-level of a module. This can be done using suitable
+analogues of the rules
+```
+(..,fix (λx.e),...) ---> fix (λx.(..,e[proj i x/x],..))
+```
+where `i` is the index in the tuple at which `fix (λx.e)` appears,
+`proj i x` selects the `i`th component from `x`, and `x` does not
+occur free elsewhere in the tuple; a corresponding rule for
+constructor applications; and
+```
+fix (λx. fix (λy.e)) ---> fix (λx. e[x/y])
+```
+Both these rules require adjusting deBruin numbers in the UPLC
+implementation.
+
+The intention here is to implement module-level recursion using a
+cyclic data-structure--the value restriction guarantees that the
+module value `Self` is not needed to compute the top-level value of
+the module, and thus there is no risk of falling into an infinite loop
+at this point. (Of course, a recursive function can loop *when it is
+called*, but constructing the function itself cannot loop because it
+must be a syntactic λ-expression). This is a *more efficient* way to
+implement recursion than the fixpoint combinators currently used in
+UPLC, and so will probably become the preferred way to implement
+recursion.
+
+#### Variation: Explicit lambdas
+
+This variation lifts some of the restrictions of the 'value scripts'

```

Comment:
Is this a variation? I thought this was just the original design

Reply:
The original design didn't require syntactic lambdas at the top-level, to bind the script arguments. This variation does. At least, that was my understanding of the original design.

Reply:
Ah I see. And the advantage is that we can rapidly pass the arguments via a pre-constructed environment?

Reply:
Yes, exactly.